File: dfs_bfs_backtracking/cpp/b1012.cpp
#include <iostream>
#include <vector>
#include <queue>
int T;
int N, M, K;

std::vector<std::pair<int, int>> directions{ {1, 0}, {-1, 0}, {0, 1}, {0, -1} };

void bfs(const int& row, const int& col, bool(&maps)[50][50]) {
	std::queue<std::pair<int, int>> check_cabbage;
	check_cabbage.push(std::make_pair(row, col));
	maps[row][col] = false;
	int temp_x, temp_y;
	int new_x, new_y;
	while (check_cabbage.size()) {
		temp_x = check_cabbage.front().first;
		temp_y = check_cabbage.front().second;
		check_cabbage.pop();
		for (std::pair<int, int>& dir : directions) {
			new_x = temp_x + dir.first;
			new_y = temp_y + dir.second;
			if (0 <= new_x && new_x < M && 0 <= new_y && new_y < N && maps[new_x][new_y]) {
				maps[new_x][new_y] = false;
				check_cabbage.push(std::make_pair(new_x, new_y));
			}
		}
	}
	return;
}

int main() {
	std::ios_base::sync_with_stdio(false);
	std::cin.tie(NULL);
	std::cin >> T;
	for (int tc = 0; tc < T; ++tc) {
		std::cin >> M >> N >> K;
		bool maps[50][50]{ false };
		std::queue<std::pair<int, int>> check;
		int x, y;
		for (int i = 0; i < K; ++i) {
			std::cin >> x >> y;
			check.push(std::make_pair(x, y));
			maps[x][y] = true;
		}
		int cnt = 0;
		while (check.size()) {
			x = check.front().first;
			y = check.front().second;
			check.pop();
			if (maps[x][y] == true) {
				cnt += 1;
				bfs(x, y, maps);
			}
		}
		std::cout << cnt << '\n';
	}
}

File: dfs_bfs_backtracking/python/b1012.py
import sys
input = sys.stdin.readline

directions = ((1, 0), (-1, 0), (0, 1), (0, -1))

def bfs(row, col):
    front = -1
    rear = 0
    queue = [0] * (R * C)
    queue[rear] = (row, col)
    maps[row][col] = False

    while front < rear:
        front += 1
        temp_r, temp_c = queue[front]
        for direction in directions:
            new_r = temp_r + direction[0]
            new_c = temp_c + direction[1]
            if 0 <= new_r < R and 0 <= new_c < C:
                if maps[new_r][new_c] == True:
                    maps[new_r][new_c] = False
                    rear += 1
                    queue[rear] = (new_r, new_c)
    return

T = int(input())
for tc in range(T):
    C, R, K = map(int, input().split())
    maps = [[False for _ in range(C)] for _ in range(R)]
    
    check = [0] * K
    for idx in range(K):
        c, r = map(int, input().split())
        maps[r][c] = True
        check[idx] = [r, c]
    cnt = 0
    for idx in range(K):
        if maps[check[idx][0]][check[idx][1]] == True:
            cnt += 1
            bfs(check[idx][0], check[idx][1])

    print(cnt)
